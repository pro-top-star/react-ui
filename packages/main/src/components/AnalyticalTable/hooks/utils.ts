export const makeTemplateColumns = (columns, resizedColumns) => {
  return columns
    .map((column) => {
      if (resizedColumns[column.id]) {
        return `${resizedColumns[column.id]}px`;
      }
      return `minmax(${column.minWidth}px, ${column.width || '1fr'})`;
    })
    .join(' ');
};
