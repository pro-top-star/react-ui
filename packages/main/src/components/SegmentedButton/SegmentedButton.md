```jsx
import { SegmentedButton } from '@ui5/webcomponents-react/lib/SegmentedButton';
```
