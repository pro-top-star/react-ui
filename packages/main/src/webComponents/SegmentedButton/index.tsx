import { withWebComponent, WithWebComponentPropTypes } from '@ui5/webcomponents-react/dist/withWebComponent';
import { Ui5CustomEvent } from '@ui5/webcomponents-react/interfaces/Ui5CustomEvent';
import { FC, ReactNode } from 'react';

import '@ui5/webcomponents/dist/SegmentedButton';

export interface SegmentedButtonPropTypes extends WithWebComponentPropTypes {
  /**
   * Defines the buttons of `SegmentedButton`.
   *
   * **Note:** Multiple buttons are allowed.
   *
   * **Note:** Use the `ToggleButton` for the intended design.
   */
  children?: ReactNode | ReactNode[];
  /**
   * Fired when the selected button changes.
   */
  onSelectionChange?: (event: Ui5CustomEvent<HTMLButtonElement, { selectedButton: ReactNode }>) => void;
}

/**
 * The `SegmentedButton` shows a group of buttons. When the user clicks or taps one of the buttons, it stays in a pressed state. It automatically resizes the buttons to fit proportionally within the component. When no width is set, the component uses the available width.
 *
 * **Note:** There can be just one selected `button` at a time.
 *
 * <ui5-link href="https://sap.github.io/ui5-webcomponents/playground/components/SegmentedButton" target="_blank">UI5 Web Components Playground</ui5-link>
 */
const SegmentedButton: FC<SegmentedButtonPropTypes> = withWebComponent<SegmentedButtonPropTypes>(
  'ui5-segmentedbutton',
  [],
  [],
  [],
  ['selection-change']
);

SegmentedButton.displayName = 'SegmentedButton';

SegmentedButton.defaultProps = {};

export { SegmentedButton };
