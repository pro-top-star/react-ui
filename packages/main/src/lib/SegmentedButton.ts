import { SegmentedButton } from '../webComponents/SegmentedButton';
import type { SegmentedButtonPropTypes } from '../webComponents/SegmentedButton';

export { SegmentedButton };
export type { SegmentedButtonPropTypes };
