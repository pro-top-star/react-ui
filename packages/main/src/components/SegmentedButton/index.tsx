import { CssSizeVariables } from '@ui5/webcomponents-react-base/lib/CssSizeVariables';
import { Event } from '@ui5/webcomponents-react-base/lib/Event';
import { StyleClassHelper } from '@ui5/webcomponents-react-base/lib/StyleClassHelper';
import { useConsolidatedRef } from '@ui5/webcomponents-react-base/lib/useConsolidatedRef';
import { usePassThroughHtmlProps } from '@ui5/webcomponents-react-base/lib/usePassThroughHtmlProps';
import React, {
  Children,
  cloneElement,
  FC,
  forwardRef,
  ReactNode,
  Ref,
  RefObject,
  useCallback,
  useEffect,
  useState
} from 'react';
import { createUseStyles } from 'react-jss';
import { CommonProps } from '../../interfaces/CommonProps';

export type SelectedKey = string | number;

export interface SegmentedButtonPropTypes extends CommonProps {
  disabled?: boolean;
  selectedKey?: SelectedKey;
  children: ReactNode | ReactNode[];
  onItemSelected?: (event: Event) => void;
}

const styles = {
  segmentedButton: {
    verticalAlign: 'top',
    position: 'relative',
    margin: '0',
    padding: CssSizeVariables.sapWcrSegmentedButtonPadding,
    border: 'none',
    whiteSpace: 'nowrap',
    display: 'inline-block',
    boxSizing: 'border-box',
    maxWidth: '100%',
    height: CssSizeVariables.sapWcrSegmentedButtonHeight,
    '&:focus': {
      outline: 'none'
    }
  }
};

const useStyles = createUseStyles(styles, { name: 'SegmentedButton' });

/**
 * <code>import { SegmentedButton } from '@ui5/webcomponents-react/lib/SegmentedButton';</code>
 */
const SegmentedButton: FC<SegmentedButtonPropTypes> = forwardRef(
  (props: SegmentedButtonPropTypes, ref: Ref<HTMLUListElement>) => {
    const { children, disabled, className, style, tooltip, slot, onItemSelected, selectedKey } = props;

    const listRef: RefObject<HTMLUListElement> = useConsolidatedRef(ref);

    const [internalSelectedKey, setSelectedKey] = useState(() => {
      if (selectedKey) return selectedKey;
      const firstChild: any = Children.toArray(children)[0];
      if (firstChild && firstChild.props) {
        return firstChild.props.id;
      }
      return null;
    });

    useEffect(() => {
      if (selectedKey) {
        setSelectedKey(selectedKey);
      }
    }, [selectedKey, setSelectedKey]);

    const classes = useStyles();

    const segmentedBtnClasses = StyleClassHelper.of(classes.segmentedButton);
    if (className) {
      segmentedBtnClasses.put(className);
    }

    const handleSegmentedButtonItemSelected = useCallback(
      (originalOnclick) => (e) => {
        const newSelectedKey = e.getParameter('selectedKey');
        if (newSelectedKey !== internalSelectedKey) {
          setSelectedKey(newSelectedKey);
          if (typeof onItemSelected === 'function') {
            onItemSelected(Event.of(null, e.getOriginalEvent(), e.getParameters()));
          }
        }
        if (typeof originalOnclick === 'function') {
          originalOnclick(e);
        }
      },
      [internalSelectedKey, setSelectedKey, onItemSelected]
    );

    useEffect(() => {
      requestAnimationFrame(() => {
        if (!listRef.current) return;
        let maxWidth = 0;
        for (let i = 0; i < listRef.current.childElementCount; i++) {
          const item = listRef.current.children.item(i) as HTMLLIElement;
          if (item.offsetWidth && item.offsetWidth > maxWidth) {
            maxWidth = item.offsetWidth;
          }
        }
        if (maxWidth < listRef.current.offsetWidth) {
          for (let i = 0; i < listRef.current.childElementCount; i++) {
            const item = listRef.current.children.item(i) as HTMLLIElement;
            if (item.getAttribute('data-has-own-width') === 'false') {
              item.style.width = `${maxWidth}px`;
            }
          }
        }
      });
    }, [children, listRef]);

    const passThroughProps = usePassThroughHtmlProps(props);

    return (
      <ul
        tabIndex={0}
        role="radiogroup"
        className={segmentedBtnClasses.toString()}
        style={style}
        ref={listRef}
        title={tooltip}
        slot={slot}
        {...passThroughProps}
      >
        {Children.toArray(children)
          .filter(Boolean)
          .map((item: any) =>
            cloneElement(item, {
              key: item.props.id,
              selected: internalSelectedKey === item.props.id,
              disabled: disabled === true ? disabled : item.props.disabled,
              onClick: handleSegmentedButtonItemSelected(item.props.onClick)
            })
          )}
      </ul>
    );
  }
);

SegmentedButton.displayName = 'SegmentedButton';

SegmentedButton.defaultProps = {
  disabled: false,
  selectedKey: '',
  onItemSelected: null
};

export { SegmentedButton };
