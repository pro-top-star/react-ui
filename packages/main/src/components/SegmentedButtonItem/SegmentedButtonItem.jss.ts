import { ContentDensity } from '@ui5/webcomponents-react/lib/ContentDensity';
import { JSSTheme } from '../../interfaces/JSSTheme';

const styles = ({ contentDensity, parameters }: JSSTheme) => ({
  segmentedButtonItem: {
    fontFamily: parameters.sapUiFontFamily,
    listStyle: 'none',
    overflow: 'hidden',
    WebkitTapHighlightColor: 'rgba(255, 255, 255, 0)',
    textAlign: 'center',
    alignContent: 'center',
    alignItems: 'center',
    cursor: 'pointer',
    verticalAlign: 'text-bottom',
    boxSizing: 'border-box',
    display: 'inline-block',
    fontSize: '0.875rem',
    '&:focus': {
      outline: `1px dotted ${parameters.sapUiContentFocusColor}`
    },
    height: contentDensity === ContentDensity.Compact ? '1.625rem' : '2.25rem',
    lineHeight: contentDensity === ContentDensity.Compact ? '1.5rem' : '2.25rem',
    textShadow: 'none',
    '&:not($iconOnly)': { paddingLeft: '0.625rem', paddingRight: '0.625rem', textOverflow: 'ellipsis' },
    '&:first-child': {
      borderTopLeftRadius: '0.25rem',
      borderBottomLeftRadius: '0.25rem'
    },
    '&:last-child': {
      borderTopRightRadius: '0.25rem',
      borderBottomRightRadius: '0.25rem',
      borderRightColor: parameters.sapUiSegmentedButtonBorderColor
    },
    color: parameters.sapUiSegmentedButtonTextColor,
    border: `1px solid ${parameters.sapUiSegmentedButtonBorderColor}`,
    borderRightColor: 'transparent',
    backgroundColor: parameters.sapUiSegmentedButtonBackground
  },
  selected: {
    background: parameters.sapUiSegmentedButtonSelectedBackground,
    color: parameters.sapUiSegmentedButtonSelectedTextColor,
    borderColor: parameters.sapUiSegmentedButtonSelectedHoverBorderColor,
    '--sapUiContentNonInteractiveIconColor': parameters.sapUiContentContrastIconColor,
    '$:active': {
      background: parameters.sapUiButtonActiveBackground,
      color: parameters.sapUiButtonActiveTextColor
    }
  },
  focusableItem: {
    '&:not($selected):hover': {
      background: parameters.sapUiSegmentedButtonHoverBackground
    }
  },
  disabled: { textShadow: 'none', cursor: 'default', opacity: '0.4' },
  iconOnly: { paddingLeft: '0.625rem', paddingRight: '0.625rem' },
  icon: {
    fontSize: '1rem',
    // width: contentDensity === ContentDensity.Compact ? '1.3rem' : '2.375rem',
    // height: contentDensity === ContentDensity.Compact ? '1.3rem' : '100%',
    // verticalAlign: contentDensity === ContentDensity.Compact ? 'top' : 'middle',
    textAlign: 'center',
    alignContent: 'center',
    display: 'inline-block',
    '&:focus': {
      outline: 'none'
    },
    '& > *': {
      verticalAlign: 'middle'
    }
  },
  withText: {
    marginRight: '0.5rem'
  }
});

export default styles;
