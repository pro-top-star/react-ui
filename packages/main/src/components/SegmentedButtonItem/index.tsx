import { Event } from '@ui5/webcomponents-react-base/lib/Event';
import { StyleClassHelper } from '@ui5/webcomponents-react-base/lib/StyleClassHelper';
import { usePassThroughHtmlProps } from '@ui5/webcomponents-react-base/lib/usePassThroughHtmlProps';
import React, { CSSProperties, FC, forwardRef, Ref, useCallback, useMemo } from 'react';
import { createUseStyles } from 'react-jss';
import { CommonProps } from '../../interfaces/CommonProps';
import { JSSTheme } from '../../interfaces/JSSTheme';
import styles from './SegmentedButtonItem.jss';

export interface SegmentedButtonItemPropTypes extends CommonProps {
  icon?: JSX.Element;
  id: string | number;
  disabled?: boolean;
  children?: string;
  width?: CSSProperties['width'];
  onClick?: (e: Event) => void;
}

const useStyles = createUseStyles<JSSTheme, keyof ReturnType<typeof styles>>(styles, { name: 'SegmentedButtonItem' });

/**
 * <code>import { SegmentedButtonItem } from '@ui5/webcomponents-react/lib/SegmentedButtonItem';</code>
 */
const SegmentedButtonItem: FC<SegmentedButtonItemPropTypes> = forwardRef(
  (props: SegmentedButtonItemPropTypes, ref: Ref<HTMLLIElement>) => {
    const { disabled, children, icon, className, style, tooltip, onClick, id, width } = props;

    const classes = useStyles();

    const iconClasses = StyleClassHelper.of(classes.icon);
    const segmentedButtonItemClasses = StyleClassHelper.of(classes.segmentedButtonItem);

    if (children && children !== '') {
      iconClasses.put(classes.withText);
    } else {
      segmentedButtonItemClasses.put(classes.iconOnly);
    }

    if (disabled) {
      segmentedButtonItemClasses.put(classes.disabled);
    } else {
      segmentedButtonItemClasses.put(classes.focusableItem);
    }

    if (props['selected']) {
      segmentedButtonItemClasses.put(classes.selected);
    }

    if (className) {
      segmentedButtonItemClasses.put(className);
    }

    const handleOnClick = useCallback(
      (e) => {
        if (!disabled && typeof onClick === 'function') {
          onClick(Event.of(null, e, { selectedKey: id }));
        }
      },
      [onClick, disabled, id]
    );

    const inlineStyles = useMemo(() => {
      if (width === undefined || width === null) {
        return style;
      }

      return {
        ...style,
        width
      };
    }, [style, width]);

    const passThroughProps = usePassThroughHtmlProps(props);

    return (
      <li
        ref={ref}
        className={segmentedButtonItemClasses.valueOf()}
        onClick={handleOnClick}
        style={inlineStyles}
        title={tooltip}
        data-has-own-width={!!width}
        {...passThroughProps}
      >
        {icon && <div className={iconClasses.valueOf()}>{icon}</div>}
        {children}
      </li>
    );
  }
);

SegmentedButtonItem.displayName = 'SegmentedButtonItem';

SegmentedButtonItem.defaultProps = {
  disabled: false
};

export { SegmentedButtonItem };
