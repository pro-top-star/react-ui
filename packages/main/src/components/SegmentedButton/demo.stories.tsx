import { action } from '@storybook/addon-actions';
import { boolean, number } from '@storybook/addon-knobs';
import { Icon } from '@ui5/webcomponents-react/lib/Icon';
import { SegmentedButton } from '@ui5/webcomponents-react/lib/SegmentedButton';
import { SegmentedButtonItem } from '@ui5/webcomponents-react/lib/SegmentedButtonItem';
import React from 'react';
import notes from './SegmentedButton.md';

export const renderStory = () => {
  return (
    <SegmentedButton
      disabled={boolean('disabled', false)}
      onItemSelected={action('onItemSelected')}
      selectedKey={number('SelectedKey', 1)}
    >
      <SegmentedButtonItem id={1} icon={<Icon name="world" />} />
      <SegmentedButtonItem id={2} icon={<Icon name="world" />}>
        Button 2
      </SegmentedButtonItem>
      <SegmentedButtonItem id={3} width="300px">
        Button 3
      </SegmentedButtonItem>
    </SegmentedButton>
  );
};
renderStory.story = {
  name: 'Default'
};

export default {
  title: 'Components / SegmentedButton',
  component: SegmentedButton,
  parameters: { notes }
};
