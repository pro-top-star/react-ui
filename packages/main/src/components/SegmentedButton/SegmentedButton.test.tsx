import { createPassThroughPropsTest, getEventFromCallback, mountThemedComponent } from '@shared/tests/utils';
import { SegmentedButton } from '@ui5/webcomponents-react/lib/SegmentedButton';
import { SegmentedButtonItem } from '@ui5/webcomponents-react/lib/SegmentedButtonItem';
import React, { cloneElement } from 'react';
import sinon from 'sinon';

describe('SegmentedButton', () => {
  test('Selection', () => {
    const callback = sinon.spy();
    const wrapper = mountThemedComponent(
      // @ts-ignore
      <SegmentedButton onItemSelected={callback} selectedKey="btn-1">
        <SegmentedButtonItem id="btn-1">Test</SegmentedButtonItem>
        <SegmentedButtonItem id="btn-2">Test</SegmentedButtonItem>
      </SegmentedButton>
    );
    wrapper
      .find('li')
      .last()
      .simulate('click');
    wrapper.update();
    expect(getEventFromCallback(callback).getParameter('selectedKey')).toEqual('btn-2');
  });

  test('Update Selection via API', () => {
    const callback = sinon.spy();

    const wrapper = mountThemedComponent(
      <SegmentedButton onItemSelected={callback}>
        <SegmentedButtonItem id="btn-1">Test</SegmentedButtonItem>
        <SegmentedButtonItem id="btn-2">Test</SegmentedButtonItem>
      </SegmentedButton>
    );
    expect(wrapper.render()).toMatchSnapshot();
    wrapper.setProps({
      children: cloneElement(wrapper.prop('children'), { selectedKey: 'btn-2' })
    });
    wrapper.update();
    expect(wrapper.render()).toMatchSnapshot();
    expect(callback.called).toBe(false);
  });

  createPassThroughPropsTest(SegmentedButton);
});
