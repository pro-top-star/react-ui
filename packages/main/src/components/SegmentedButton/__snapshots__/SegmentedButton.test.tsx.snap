// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SegmentedButton Update Selection via API 1`] = `
<ul
  class="SegmentedButton--segmentedButton-"
  role="radiogroup"
  tabindex="0"
>
  <li
    class="SegmentedButtonItem--segmentedButtonItem- SegmentedButtonItem--focusableItem- SegmentedButtonItem--selected-"
    data-has-own-width="false"
    id="btn-1"
  >
    Test
  </li>
  <li
    class="SegmentedButtonItem--segmentedButtonItem- SegmentedButtonItem--focusableItem-"
    data-has-own-width="false"
    id="btn-2"
  >
    Test
  </li>
</ul>
`;

exports[`SegmentedButton Update Selection via API 2`] = `
<ul
  class="SegmentedButton--segmentedButton-"
  role="radiogroup"
  tabindex="0"
>
  <li
    class="SegmentedButtonItem--segmentedButtonItem- SegmentedButtonItem--focusableItem-"
    data-has-own-width="false"
    id="btn-1"
  >
    Test
  </li>
  <li
    class="SegmentedButtonItem--segmentedButtonItem- SegmentedButtonItem--focusableItem- SegmentedButtonItem--selected-"
    data-has-own-width="false"
    id="btn-2"
  >
    Test
  </li>
</ul>
`;
